### PR TITLE
Rework Support Ops to use TensorList

### PR DESCRIPTION
**Pull request by klecki**
Extend OpSpec GetArgument
Add test to verify how ArgumentInputs can be accessed

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

#### Why we need this PR?
- Change the output type of SupportOps to TensorList so they can be replaced by regular CPU/GPU ops in the future.

#### What happened in this PR?
 - Explain solution of the problem, new feature added.
Output of SupportOp is changed to TensorList. It also allows for arbitrary shaped batches to be used as Argument Inputs.
 - What was changed, added, removed?
OpSpec code for obtaining tensor arguments was adjusted.
 - What is most important part that reviewers should focus on?
N/A
 - Was this PR tested? How?
A test was added to check the constraints on the Tensor Argument shape and access to Argument Inputs. CI will run L3.
 - Were docs and examples updated, if necessary?

**JIRA TASK**: [DALI-1044]

### CODE REVIEW

File: dali/pipeline/operators/op_spec_test.cc [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+DALI_SCHEMA(SupportAccessingOp)
+  .DocStr("SupportAccessingOp")
+  .NumInput(0)
+  .NumOutput(1)
+  .AddOptionalArg("arg0", "no-doc", 42, true)
+  .AddOptionalArg("arg1", "no-doc", 42.f, true)
+  .AddOptionalArg("arg2", "no-doc", 42, true)
+  .AddOptionalArg("arg3", "no-doc", 42, true);
+
+TEST(ArgumentInputTest, OpSpecAccess) {
```
Comment (awolant) [inline dali/pipeline/operators/op_spec_test.cc]: Very cool way to test this :) Maybe add a bit of comment as it is not "standard" test.

File: dali/pipeline/operators/op_spec.h [hunk 2]
```diff
@@ -445,7 +445,36 @@ inline T OpSpec::GetArgumentImpl(
...
+            "Unexpected shape of argument \"" + name + "\". Expected batch of " +
+                std::to_string(batch_size) + " tensors of shape {1}, got " +
+                std::to_string(shape.num_samples()) + " samples of " +
+                std::to_string(shape.sample_dim()) +
+                "-dim tensors. Alternatively 1 tensor of shape {" + std::to_string(batch_size) +
+                "} can be passed. To access argument inputs where samples are not scalars "
+                "use ArgumentWorkspace::ArgumentInput method directly.");
+      }
+    };
+    check_shape(value.shape(), GetArgument<int>("batch_size"), name);
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]: Isn't there a way to get batch size without referring to it by name (using string comparison)?

File: dali/pipeline/operators/op_spec.h [hunk 3]
```diff
@@ -474,6 +503,9 @@ inline bool OpSpec::TryGetArgumentImpl(
     if (ws == nullptr)
       return false;
     const auto& value = ws->ArgumentInput(name);
+    if (value.shape() != kernels::uniform_list_shape(GetArgument<int>("batch_size"), {1})) {
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]: Sorry, we can't have it - this is FAR too expensive for a simple check, executed on a per-sample basis.

File: dali/pipeline/operators/op_spec_test.cc [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+#include "dali/pipeline/data/buffer.h"
+#include "dali/pipeline/data/tensor.h"
+#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/workspace/workspace.h"
+#include "dali/test/dali_test.h"
+
+
+namespace dali {
+
+class SupportGeneratingOp : public Operator<SupportBackend> {
```
Comment (mzient) [inline dali/pipeline/operators/op_spec_test.cc]: I think the word "test" should appear somewhere in the name - after all, this operator is globally visible to all operators in dali_test and it'd be nice if it was clearly distinguishable from normal operators exposed by dali.so.

Comment (mzient) [reply dali/pipeline/operators/op_spec_test.cc]: Suggested alternatives:
`TestSupportOp_Generator`
`TestSupportOp_Consumer`

File: dali/pipeline/operators/op_spec_test.cc [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+DALI_REGISTER_OPERATOR(SupportGeneratingOp, SupportGeneratingOp, Support);
+
+DALI_SCHEMA(SupportGeneratingOp)
+  .DocStr("SupportGeneratingOp")
+  .NumInput(0)
+  .NumOutput(4);
+
+
+class SupportAccessingOp : public Operator<CPUBackend> {
```
Comment (mzient) [inline dali/pipeline/operators/op_spec_test.cc]: Naming - see above.

File: dali/pipeline/operators/op_spec.h (hunk 2 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec.h]: It's passed this way to operators, and the OperatorBase stores it as a member. But for OpSpec it's only an internal argument, I don't know about any other way.

File: dali/pipeline/operators/op_spec.h (hunk 3 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec.h]: I reused the logic from GetArgument.

File: dali/pipeline/operators/op_spec_test.cc (hunk 4 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec_test.cc]: Done

File: dali/pipeline/operators/op_spec_test.cc (hunk 5 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec_test.cc]: Done

File: dali/pipeline/operators/op_spec_test.cc (hunk 1 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec_test.cc]: Done

**Comment by klecki** [comment]
!build

File: dali/pipeline/operators/op_spec.h [hunk 6]
```diff
@@ -416,6 +416,50 @@ class DLL_PUBLIC OpSpec {
...
+   * @brief Check if the ArgumentInput of given shape can be used with GetArgument(),
+   *        representing a batch of scalars
+   *
+   * @argument should_throw whether this function should throw an error if the shape doesn't match
+   * @return true iff the shape is allowed to be used as Argument
+   */
+  bool CheckArgumentShape(const kernels::TensorListShape<> &shape, int batch_size,
+                          const std::string &name, bool should_throw = false) const {
+    if (shape.num_samples() == 1) {
+      // TODO(klecki): 1 sample version will be of no use after the switch to CPU ops unless we
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]: Then maybe we should do something about 0-D tensors...

File: dali/pipeline/operators/displacement/warp_param_provider.h [hunk 7]
```diff
@@ -229,19 +229,29 @@ class WarpParamProvider : public InterpTypeProvider, public BorderTypeProvider<B
...
     const int N = num_samples_;
+
+    DALI_ENFORCE(is_uniform(shape), "Output sizes must be passed as uniform Tensor List.");
     DALI_ENFORCE(
-      tv.shape == kernels::TensorShape<>(N, spatial_ndim) ||
-      tv.shape == kernels::TensorShape<>(N * spatial_ndim),
-      "Output sizes must either be a flat array of size num_samples*dim "
-      "or a 2D tensor of size num_samples x dim");
+        (shape.num_samples() == N && shape[0] == kernels::TensorShape<>(spatial_ndim)) ||
+            (shape.num_samples() == 1 && (shape[0] == kernels::TensorShape<>(N, spatial_ndim) ||
```
Comment (mzient) [inline dali/pipeline/operators/displacement/warp_param_provider.h]: Indentation doesn't match parentheses.

File: dali/pipeline/operators/displacement/warp_param_provider.h [hunk 8]
```diff
@@ -229,19 +229,29 @@ class WarpParamProvider : public InterpTypeProvider, public BorderTypeProvider<B
...
+                                          shape[0] == kernels::TensorShape<>(N * spatial_ndim))),
+        "Output sizes must either be a batch of `dim`-sized tensors, flat array of size "
+        "num_samples*dim or a one 2-dim tensor of shape {num_samples, dim}.");
```
Comment (mzient) [inline dali/pipeline/operators/displacement/warp_param_provider.h]: Why have you replaced 2D with 2-dim? It's confusing in the presence of `dim` here...
```suggestion
        "num_samples*dim or one 2D tensor of shape {num_samples, dim}.");
```

File: dali/pipeline/operators/op_spec.h [hunk 9]
```diff
@@ -416,6 +416,50 @@ class DLL_PUBLIC OpSpec {
...
+                " tensors of shape {1} are expected. To access argument inputs where samples are "
+                "not scalars use ArgumentWorkspace::ArgumentInput method directly.");
+      } else if (!is_one_sample_with_batch) {
+        return false;
+      }
+    } else {
+      bool is_batch_of_scalars =
+          shape.num_samples() == batch_size && shape.sample_dim() == 1 && shape[0][0] == 1;
+      if (should_throw) {
+        DALI_ENFORCE(is_batch_of_scalars,
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]: :+1: For a meaningful condition in DALI_ENFORCE.

File: dali/pipeline/operators/op_spec.h [hunk 10]
```diff
@@ -416,6 +416,50 @@ class DLL_PUBLIC OpSpec {
...
+            "Unexpected shape of argument \"" + name + "\". Expected batch of " +
+                std::to_string(batch_size) + " tensors of shape {1}, got " +
+                std::to_string(shape.num_samples()) + " samples of " +
+                std::to_string(shape.sample_dim()) +
+                "-dim tensors. Alternatively 1 tensor of shape {" + std::to_string(batch_size) +
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]:
```suggestion
                "-D tensors. Alternatively, a single 1-D tensor with " + std::to_string(batch_size) + " elements "
```

File: dali/pipeline/operators/op_spec.h [hunk 11]
```diff
@@ -416,6 +416,50 @@ class DLL_PUBLIC OpSpec {
...
+   * @return true iff the shape is allowed to be used as Argument
+   */
+  bool CheckArgumentShape(const kernels::TensorListShape<> &shape, int batch_size,
+                          const std::string &name, bool should_throw = false) const {
+    if (shape.num_samples() == 1) {
+      // TODO(klecki): 1 sample version will be of no use after the switch to CPU ops unless we
+      // generalize accepted batch sizes throughout the pipeline
+      bool is_one_sample_with_batch = shape[0] == kernels::TensorShape<>(batch_size);
+      if (should_throw) {
+        DALI_ENFORCE(is_one_sample_with_batch,
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]: :+1: for meaningful condition in DALI_ENFORCE

File: dali/pipeline/operators/op_spec.h [hunk 12]
```diff
@@ -416,6 +416,50 @@ class DLL_PUBLIC OpSpec {
...
+            "Unexpected shape of argument \"" + name +
+                "\". If only one tensor is passed, it should have a shape equal to {" +
+                std::to_string(batch_size) +
+                "}.  When accessing arguments as scalars 1 tensor of shape {" +
+                std::to_string(batch_size) + "} or " + std::to_string(batch_size) +
+                " tensors of shape {1} are expected. To access argument inputs where samples are "
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]: This `{1}` looks weird - made me look for some kind of formatting library.

File: dali/pipeline/operators/op_spec_test.cc [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+  }
+};
+
+DALI_REGISTER_OPERATOR(TestArgumentInput_Consumer, TestArgumentInput_Consumer, CPU);
+
+DALI_SCHEMA(TestArgumentInput_Consumer)
+    .DocStr("TestArgumentInput_Consumer")
+    .NumInput(0)
+    .NumOutput(1)
+    .AddOptionalArg("arg0", "no-doc", 42, true)
```
Comment (mzient) [inline dali/pipeline/operators/op_spec_test.cc]: Using same value in all arguments makes it possible to miss a bug when all arguments have same default value.

File: dali/pipeline/operators/op_spec.h [hunk 14]
```diff
@@ -445,7 +490,12 @@ inline T OpSpec::GetArgumentImpl(
   // Search for the argument in tensor arguments first
   if (this->HasTensorArgument(name)) {
     DALI_ENFORCE(ws != nullptr, "Tensor value is unexpected for argument \"" + name + "\".");
-    const auto& value = ws->ArgumentInput(name);
+    const auto &value = ws->ArgumentInput(name);
+    DALI_ENFORCE(kernels::is_uniform(value.shape()),
+                 "Arguments should be passed as uniform TensorLists. Argument \"" + name +
```
Comment (mzient) [inline dali/pipeline/operators/op_spec.h]: Aren't non-uniform argument inputs legal when obtained via `ArgumentInput`? This error suggests they aren't.

Comment (mzient) [reply dali/pipeline/operators/op_spec.h]: Also, the check for uniformity could be inside CheckArgumentShape - now it's not checked in TryGetArgumentImpl.

File: dali/pipeline/operators/displacement/warp_param_provider.h (hunk 7 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/warp_param_provider.h]: But it matches the indent rules about expression continuation from clang-format.

File: dali/pipeline/operators/displacement/warp_param_provider.h (hunk 8 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/warp_param_provider.h]: Done

File: dali/pipeline/operators/op_spec.h (hunk 10 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec.h]: Done

File: dali/pipeline/operators/op_spec.h (hunk 14 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec.h]: Still, this check is not used with ArgumentInput but only with Arguments.

File: dali/pipeline/operators/op_spec.h (hunk 12 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec.h]: But that's the actual shape.

File: dali/pipeline/operators/op_spec_test.cc (hunk 13 above)
Comment (klecki) [reply dali/pipeline/operators/op_spec_test.cc]: This is basically only used to infer the type, I am not accessing the default values, as in the test, the argument inputs are always provided.

**Comment by klecki** [comment]
!build

File: dali/pipeline/operators/displacement/warp_param_provider.h (hunk 7 above)
Comment (mzient) [reply dali/pipeline/operators/displacement/warp_param_provider.h]: So what? Clang-format is not omniscient and, apparently, in this case failed miserably at capturing the sense of the expression. It makes it look like the contents of this line are inside the parenthesis opened above - and it's not the case.

File: dali/pipeline/operators/op_spec.h (hunk 12 above)
Comment (mzient) [reply dali/pipeline/operators/op_spec.h]: I know - this just looks weird in the code. BTW - when we print TensorShape, we don't use braces - they are used in TensorListShape only; tensor shape would be printed as just `1`

**Comment by klecki** [comment]
!build

**Pull request merged**